README.md
# flutter_image_slideshow

A simple image slideshow widget.
Mainly intended for image widget, but other widgets can also be used.

<image src="https://raw.github.com/wiki/edasandesu/flutter_image_slideshow/images/slideshow.gif" width=270>

## Getting Started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  flutter:
    sdk:
  flutter_image_slideshow: any
```

In your library add the following import:

```dart
  import 'package:flutter_image_slideshow/flutter_image_slideshow.dart';
```

## Usage

These are The ImageSlideshow properties.

```dart
/// The widgets to display in the [ImageSlideshow].
///
/// Mainly intended for image widget, but other widgets can also be used.
List<Widget> children

/// Width of the [ImageSlideshow].
double width

/// Height of the [ImageSlideshow].
double height

/// The page to show when first creating the [ImageSlideshow].
int initialPage

/// The color to paint the indicator.
Color indicatorColor

/// The color to paint behind th indicator.
Color indicatorBackgroundColor

/// Called whenever the page in the center of the viewport changes.
ValueChanged<int> onPageChanged

/// Auto scroll interval.
///
/// Do not auto scroll with null or 0.
int autoPlayInterval

/// Loops back to first slide.
bool isLoop

/// Radius of circle indicator.
double indicatorRadius

/// Disable page changes by the user.
bool disableUserScrolling
```

## Example

Create the ImageSlideshow widget.
And pass the parameters.

```dart
import 'package:flutter/material.dart';
import 'package:flutter_image_slideshow_test/flutter_image_slideshow.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: Text('Demo'),
        ),
        body: ImageSlideshow(

          /// Width of the [ImageSlideshow].
          width: double.infinity,

          /// Height of the [ImageSlideshow].
          height: 200,

          /// The page to show when first creating the [ImageSlideshow].
          initialPage: 0,

          /// The color to paint the indicator.
          indicatorColor: Colors.blue,

          /// The color to paint behind th indicator.
          indicatorBackgroundColor: Colors.grey,

          /// The widgets to display in the [ImageSlideshow].
          /// Add the sample image file into the images folder
          children: [
            Image.asset(
              'images/sample_image_1.jpg',
              fit: BoxFit.cover,
            ),
            Image.asset(
              'images/sample_image_2.jpg',
              fit: BoxFit.cover,
            ),
            Image.asset(
              'images/sample_image_3.jpg',
              fit: BoxFit.cover,
            ),
          ],

          /// Called whenever the page in the center of the viewport changes.
          onPageChanged: (value) {
            print('Page changed: $value');
          },

          /// Auto scroll interval.
          /// Do not auto scroll with null or 0.
          autoPlayInterval: 3000,

          /// Loops back to first slide.
          isLoop: true,
        ),
      ),
    );
  }
}
```


CHANGELOG.md
## abhlabs pull request Jul 25, 2023
- Added "resumeAutoPlay" function
- Added "pauseAutoPlay" function which simply calls "stopAutoPlay"
- Revised sample program to demostrate the use of a GlobalKey to access the 
    flutter_image_slideshow state, and to call the public functions.

## 0.1.6 - Jul 2, 2023

- Add customize indicator padding. [#18](https://github.com/edasandesu/flutter_image_slideshow/pull/18)
- Fix the issue where Timer is not being disposed [#14](https://github.com/edasandesu/flutter_image_slideshow/issues/14)

## 0.1.5 - Mar 28, 2023

- Add option for disabling manual page changes by the user [#16](https://github.com/edasandesu/flutter_image_slideshow/pull/16)

## 0.1.4 - Aug 15, 2022

Add customize size indicator.

## 0.1.3 - Aug 12, 2022

Make "ImageSlideshowState" public.
Add "goToPage" and "stopAutoPlay"

## 0.1.2 - Oct 3, 2021.

Add the "scrollBehavior".

## 0.1.1 - July 27, 2021.

Add an "isLoop" variable.

## 0.1.0 - May 6, 2021.

Migrate to null safety.

## 0.0.4 - Apr 12, 2021.

Add processing to check clients before animateToPage.

## 0.0.3 - Feb 28, 2021.

Add an "AutoPlay" feature.

## 0.0.2+4 - Feb 15, 2021.

Delete log.

## 0.0.2+3 - Feb 15, 2021.

Fix when onPageChanged is null.

## 0.0.2+2 - Jan 28, 2021.

Update README.md

## 0.0.2+1 - Nov 16, 2020.

Update README.md

## 0.0.2 - Nov 16, 2020.

Now with onPageChanged customisations you can also make many other processing.

## 0.0.1 - Sep 18, 2020.

Initial Version of the library.


test/flutter_image_slideshow_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_image_slideshow/flutter_image_slideshow.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/src/indicator.dart
import 'package:flutter/material.dart';

class Indicator extends StatelessWidget {
  const Indicator({
    Key? key,
    required this.count,
    required this.currentIndex,
    this.activeColor,
    this.backgroundColor,
    required this.padding,
    required this.radius,
  }) : super(key: key);

  final int count;
  final int currentIndex;
  final Color? activeColor;
  final Color? backgroundColor;
  final double padding;
  final double radius;

  @override
  Widget build(BuildContext context) {
    return Wrap(
      spacing: padding,
      runSpacing: padding,
      alignment: WrapAlignment.center,
      children: List.generate(
        count,
        (index) {
          return CircleAvatar(
            radius: radius,
            backgroundColor:
                currentIndex == index ? activeColor : backgroundColor,
          );
        },
      ),
    );
  }
}


lib/src/image_slideshow.dart
import 'dart:async';

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter_image_slideshow/src/indicator.dart';

class ImageSlideshow extends StatefulWidget {
  const ImageSlideshow({
    Key? key,
    required this.children,
    this.width = double.infinity,
    this.height = 200,
    this.initialPage = 0,
    this.indicatorColor,
    this.indicatorBackgroundColor = Colors.grey,
    this.onPageChanged,
    this.autoPlayInterval,
    this.isLoop = false,
    this.indicatorRadius = 3,
    this.indicatorPadding = 4,
    this.indicatorBottomPadding = 10,
    this.disableUserScrolling = false,
  }) : super(key: key);

  /// The widgets to display in the [ImageSlideshow].
  ///
  /// Mainly intended for image widget, but other widgets can also be used.
  final List<Widget> children;

  /// Width of the [ImageSlideshow].
  final double width;

  /// Height of the [ImageSlideshow].
  final double height;

  /// The page to show when first creating the [ImageSlideshow].
  final int initialPage;

  /// The color to paint the indicator.
  final Color? indicatorColor;

  /// The color to paint behind th indicator.
  final Color? indicatorBackgroundColor;

  /// Called whenever the page in the center of the viewport changes.
  final ValueChanged<int>? onPageChanged;

  /// Auto scroll interval.
  ///
  /// Do not auto scroll when you enter null or 0.
  final int? autoPlayInterval;

  /// Loops back to first slide.
  final bool isLoop;

  /// Radius of CircleIndicator.
  final double indicatorRadius;

  /// Padding of CircleIndicator.
  final double indicatorPadding;

  /// BottomPadding to Indicator.
  final double indicatorBottomPadding;

  /// Disable page changes by the user.
  final bool disableUserScrolling;

  @override
  ImageSlideshowState createState() => ImageSlideshowState();
}

class ImageSlideshowState extends State<ImageSlideshow> {
  late final ValueNotifier<int> _currentPageNotifier;
  late final PageController _pageController;
  late final ScrollBehavior _scrollBehavior;
  Timer? _timer;

  void _onPageChanged(int index) {
    _currentPageNotifier.value = index;
    if (widget.onPageChanged != null) {
      final correctIndex = index % widget.children.length;
      widget.onPageChanged!(correctIndex);
    }
  }

  void _autoPlayTimerStart() {
    _timer?.cancel();
    _timer = Timer.periodic(
      Duration(milliseconds: widget.autoPlayInterval!),
      (timer) {
        int nextPage;
        if (widget.isLoop) {
          nextPage = _currentPageNotifier.value + 1;
        } else {
          if (_currentPageNotifier.value < widget.children.length - 1) {
            nextPage = _currentPageNotifier.value + 1;
          } else {
            return;
          }
        }

        goToPage(nextPage);
      },
    );
  }

  void goToPage(int index) {
    if (_pageController.hasClients) {
      _pageController.animateToPage(
        index,
        duration: const Duration(milliseconds: 350),
        curve: Curves.easeIn,
      );
    }
  }

  void stopAutoPlay() {
    _timer?.cancel();
  }


  void pauseAutoPlay () {
    stopAutoPlay();
  }

  void resumeAutoPlay () {
    final page = _currentPageNotifier.value + 1;
    goToPage (page);
    _autoPlayTimerStart();
  }

  @override
  void initState() {
    _scrollBehavior = widget.disableUserScrolling
        ? const ScrollBehavior().copyWith(
            scrollbars: false,
            dragDevices: {},
          )
        : const ScrollBehavior().copyWith(
            scrollbars: false,
            dragDevices: {
              PointerDeviceKind.touch,
              PointerDeviceKind.mouse,
            },
          );

    _pageController = PageController(
      initialPage: widget.initialPage,
    );

    _currentPageNotifier = ValueNotifier(widget.initialPage);

    if (widget.autoPlayInterval != null && widget.autoPlayInterval != 0) {
      _autoPlayTimerStart();
    }
    super.initState();
  }

  @override
  void dispose() {
    _pageController.dispose();
    _currentPageNotifier.dispose();
    _timer?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: widget.width,
      height: widget.height,
      child: Stack(
        children: [
          PageView.builder(
            scrollBehavior: _scrollBehavior,
            onPageChanged: _onPageChanged,
            itemCount: widget.isLoop ? null : widget.children.length,
            controller: _pageController,
            itemBuilder: (context, index) {
              final correctIndex = index % widget.children.length;
              return widget.children[correctIndex];
            },
          ),
          Positioned(
            left: 0,
            right: 0,
            bottom: widget.indicatorBottomPadding,
            child: ValueListenableBuilder<int>(
              valueListenable: _currentPageNotifier,
              builder: (context, value, child) {
                return Indicator(
                  count: widget.children.length,
                  currentIndex: value % widget.children.length,
                  activeColor: widget.indicatorColor,
                  backgroundColor: widget.indicatorBackgroundColor,
                  radius: widget.indicatorRadius,
                  padding: widget.indicatorPadding,
                );
              },
            ),
          ),
        ],
      ),
    );
  }
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_image_slideshow/flutter_image_slideshow.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {

  final appGlobalKey = GlobalKey<ImageSlideshowState>(); 
  bool isPaused = false;
  String pauseResume = 'Pause';

  void doPauseResume () {
    final state = appGlobalKey.currentState;
    if (isPaused) {
      state!.resumeAutoPlay();
      setState ((){
        isPaused = false;
        pauseResume = 'Pause';
      });
    } else {
      state!.stopAutoPlay();
      setState (() {
        isPaused = true;
        pauseResume = 'Resume';
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Demo'),
        ),
        body: Column(
          children: [
            ImageSlideshow(
              key: appGlobalKey,
              indicatorColor: Colors.blue,
              onPageChanged: (value) {
                debugPrint('Page changed: $value');
              },
              autoPlayInterval: 3000,
              isLoop: true,
              children: [

// Lorem Picsum:  https://picsum.photos/
// Royalty free placeholder photos
                Image.network ('https://picsum.photos/id/169/450/275',
                  fit: BoxFit.cover,
                ),
                Image.network ('https://picsum.photos/id/237/450/275',
                  fit: BoxFit.cover,
                ),
                Image.network ('https://picsum.photos/id/659/450/275',
                  fit: BoxFit.cover,
                ),
              ],
            ),

            const SizedBox (height: 30,),

            ElevatedButton (
              style: ElevatedButton.
                styleFrom(minimumSize: const Size (80, 44),), 
              onPressed: doPauseResume, 
              child: Text (pauseResume, style: const TextStyle (fontSize: 16)),
            ),

          ],
        ),
      ),
    );
  }
}
